### PR TITLE
Enhance openssl tracing to fallback to function pointer addrs when dlsym fails

### PR DESCRIPTION
**Pull request by ddelnano**
This is the next step for #407.

The netty-tcnative shared library does not use a dynamic `OpenSSL_num_version` symbol. This means the symbol ends up in the symtab rather than dynsym. The former is private and incompatible with dlsym, while the latter is public and is supported by the existing openssl tracing. See this [slack thread](https://pixie-community.slack.com/archives/C03461U99S6/p1646408481645759) for more background.

This PR needs some work, but it's far enough along that I wanted  to seek feedback on my approach

## Testing
- The `openssl_trace_bpf_test` tests cases pass
  - [x] nginx openssl 1.1.1
  - [x] nginx openssl 1.1.0
  - [x] Node 12.3.1
  - [x] Node 14.18.1

Assuming this approach is how we want to proceed, the following will need to be addressed

## Todo
- [x] Parameterized `openssl_trace_bpf_test` to test the `dlsym` and `RawFptrManager` implementations of openssl tracing
- [x] ~Write unit tests for `RawFptrManager`~ -- I think our bpf test is sufficient for now and the scaffolding required to test this would be quite a bit of work so I'm holding off on this for now
- [x] ~Clean up the code added to ProcParser (much of it is duplicated with another function)~ -- not sure this is necessary
- [x] ~Make `ProcessMap` more efficient (see TODO comment about using octal numbers)~ -- wasn't sure if doing so would be more efficient. Open to any feedback you have here
- [x] Fix remaining `openssl_trace_bpf_test` test cases. They currently cause seg faults. Each test case passes individually, but when more than one case is enabled it results in seg faults ([pastebin](https://pastebin.com/Bju3TcSB))

### CODE REVIEW

File: src/stirling/obj_tools/elf_reader.cc
Comment (ddelnano) [inline src/stirling/obj_tools/elf_reader.cc:170]: Were the segments skipped intentionally or was this not a need before?

File: src/stirling/source_connectors/socket_tracer/uprobe_symaddrs.h
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/uprobe_symaddrs.h:47]: Please let me know if how I'm handling these pointers is dangerous. I'm assuming that since the caller is using a `std::unique_ptr` that I won't have to assume any responsibility of managing its lifecycle. This is an area where I'm not confident what the correct thing to do is.

Comment (yzhao1012) [reply src/stirling/source_connectors/socket_tracer/uprobe_symaddrs.h:47]: Should be fine.

We have a few other places use ElfReader* ProcParser* in function as input argument to be used inside.

However, based on what I read, I think a cleaner way is to let RawFptrManager's ctor accept ProcessMap, and instead letter caller of RawFptrManager to call ProcParser to get the ProcessMap. That should not affect the overall flow of RawFptrManager.

WDYT?

Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/uprobe_symaddrs.h:47]: I actually considered this, but I implemented it this way for the following reasons.

When `dlsym` works, we won't perform any additional work (aside from constructing the object). Doing it as you described means we will parse `/proc/$PID/maps` and `dlopen` the library an extra time, regardless if we need to fall back on this implementation. We could defer the `ProcParser::GetExecutableMapEntry` call until we need it, but then the `ProcParser` will need to be passed much further in the `uprobe_symaddrs` code. So I thought this separation was cleaner in order to defer retrieving the ProcessMap.

In addition, if the RawFptrManager's ctor accepted a ProcessMap it means that the caller could construct an object with a mismatched `lib_path` and `process_map`. See the example below for what I mean

```
# ctr definition
RawFptrManager(obj_tools::ElfReader* elf_reader, ProcessMap process_map, std::string lib_path);

std::string libpath = " ... ";
auto map = proc_parser_->GetExecutableMapEntry(pid, "not the lib path above");

// dlopen will get called on an incorrect binary
RawFptrManager(elf_reader, map, libpath);
```

Let me know what you think on that justification.

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:42]: Ideally the test suite would be able to run both cases for all of the tests. Any suggestions on how that could be accomplished?

File: src/common/system/proc_parser.h
Comment (ddelnano) [reply src/common/system/proc_parser.h:359]: This is resolved in a later commit.

File: src/common/system/proc_parser.h
Comment (ddelnano) [reply src/common/system/proc_parser.h]: Done, this was left over while I was working on fixing the more pressing segfault issues

File: src/common/system/proc_parser.h
Comment (ddelnano) [reply src/common/system/proc_parser.h]: Fixed

File: src/common/system/proc_parser.h
Comment (ddelnano) [reply src/common/system/proc_parser.h]: I've removed the `= "";` and this TODO since I wasn't sure if changing the `GetMapEntries` code would be more performant if it was backed by an 8 bit field.

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (oazizi000) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:42]: @ddelnano, a couple of pointers:
1) You can use `INSTANTIATE_TEST_SUITE_P` which is part of gtest to pass in parameters to a test. You can find plenty of examples in Stirling if you grep around.
2) You can use PL_SET_FOR_SCOPE for assigning flags in tests. Since flags are global, it helps to avoid unintended consequences when changing a flag (i.e. so the flag only applies for the test case in questions).

File: src/stirling/obj_tools/elf_reader.cc
Comment (oazizi000) [reply src/stirling/obj_tools/elf_reader.cc:170]: May have been unintentional. Can you say more about what you're looking for in he segments?

Comment (ddelnano) [reply src/stirling/obj_tools/elf_reader.cc:170]: Since segments are what end up memory mapped, we need to find the segment offset where the `.text` section is located to find the function's memory address.

Here is an example `libcrypto.so.1.1` file and the corresponding segments:

```
vagrant@vagrant:/vagrant$ readelf -lW libcrypto.so.1.1

Elf file type is DYN (Shared object file)
Entry point 0x79000
There are 9 program headers, starting at offset 64

Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  LOAD           0x000000 0x0000000000000000 0x0000000000000000 0x077930 0x077930 R   0x1000
  LOAD           0x078000 0x0000000000078000 0x0000000000078000 0x19a80d 0x19a80d R E 0x1000
  LOAD           0x213000 0x0000000000213000 0x0000000000213000 0x090420 0x090420 R   0x1000
  LOAD           0x2a3ab0 0x00000000002a4ab0 0x00000000002a4ab0 0x02d0c0 0x0314f0 RW  0x1000
  DYNAMIC        0x2ce4c8 0x00000000002cf4c8 0x00000000002cf4c8 0x000230 0x000230 RW  0x8
  NOTE           0x000238 0x0000000000000238 0x0000000000000238 0x000024 0x000024 R   0x4
  GNU_EH_FRAME   0x256c58 0x0000000000256c58 0x0000000000256c58 0x00c904 0x00c904 R   0x4
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x10
  GNU_RELRO      0x2a3ab0 0x00000000002a4ab0 0x00000000002a4ab0 0x02b550 0x02b550 R   0x1

 Section to Segment mapping:
  Segment Sections...
   00     .note.gnu.build-id .gnu.hash .dynsym .dynstr .gnu.version .gnu.version_d .gnu.version_r .rela.dyn .rela.plt
   01     .init .plt .plt.got .text .fini
   02     .rodata .eh_frame_hdr .eh_frame
   03     .init_array .fini_array .data.rel.ro .dynamic .got .data .bss
   04     .dynamic
   05     .note.gnu.build-id
   06     .eh_frame_hdr
   07
   08     .init_array .fini_array .data.rel.ro .dynamic .got
```

The 0x078000 offset that corresponds to segment 01 needs to be determined in order to find a function's memory address properly. You can see the full calculation [here](https://github.com/pixie-io/pixie/pull/419/files#diff-e26509bdd023bc2adee54cfb2c8c0c039400d90c6c8e7a2a12edf454278d36ecR518-R521)).

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:42]: 1. I did notice the `INSTANTIATE_TEST_SUITE_P` examples, but the openssl_trace_bpf_test code uses `TYPED_TEST`s. I wasn't sure how to incorporate the parameter aspects when typed tests are in use
2. Thanks for that tip. `PL_SET_FOR_SCOPE` is exactly what I was looking for.

File: src/stirling/source_connectors/socket_tracer/uprobe_symaddrs.h
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/uprobe_symaddrs.h]: @yzhao1012 just checking in to see if you saw my previous comment.

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:42]: This [Stackoverflow post](type-parameterized tests and value-parameterized tests) post mentions that the two features are not complementary and links to a google group [discussion](https://groups.google.com/g/googletestframework/c/3d8H5J5Isnw) with the maintainers. And given the copy/move issues with the test cases due to the underlying ContainerRunner instances, I think we make the typed tests work.

I've crated a simpler example (copied below). I need to figure out how to declare each `TYPED_TEST` multiple times still, which is what I'm trying to accomplish with the `OPENSSL_TYPED_TEST` macro. Please let me know your thoughts on that approach or any other ideas you have.

```
#include <gtest/gtest.h>

class ContainerRunner {
  public:
    void Run() {
    }
};

class ContainerOne : public ContainerRunner { };
class ContainerTwo : public ContainerRunner { };

template <typename TServerContainer, bool TForceFptr>
class BaseOpenSSLTraceTest : public ::testing::Test {
  public:
    BaseOpenSSLTraceTest() {
      server_.Run();
    }
    TServerContainer server_;
    bool force_fptr_ = TForceFptr;

};

template <typename T>
using OpenSSLTraceTest = BaseOpenSSLTraceTest<T, false>;

typedef ::testing::Types<ContainerOne, ContainerTwo>
  OpenSSLServerImplementations;

// Old way that works
TYPED_TEST_SUITE(OpenSSLTraceTest, OpenSSLServerImplementations);

#define OPENSSL_TYPED_TEST(TestName, TestCase, CodeBlock) TYPED_TEST(TestName, TestCase) CodeBlock

OPENSSL_TYPED_TEST(OpenSSLTraceTest, TestOne, {
  printf("TestOne test");
})

```

Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc]: @oazizi000 please see how I implemented this in https://github.com/pixie-io/pixie/pull/419/commits/6f0fbf0f127fcd0fb854a015bd5f5cb669e8cd4d

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:95]: Since this value is read during `SocketTraceBPFTest::SetUp()`, we can't use `PL_SET_FOR_SCOPE` as discussed [here](https://github.com/pixie-io/pixie/pull/419#discussion_r840967938).

File: src/common/system/proc_parser.cc
Comment (oazizi000) [inline src/common/system/proc_parser.cc:786]: Looks like we're both using `ReadFileToString` and `ifs.open()`. Isn't that redundant?

File: src/common/system/proc_parser.cc
Comment (oazizi000) [inline src/common/system/proc_parser.cc:805]: This looks like code we have in `ParseProcPIDSMaps()`. Seems like an opportunity to refactor and share?

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (oazizi000) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:95]: Makes sense

File: src/stirling/obj_tools/elf_reader.cc
Comment (oazizi000) [reply src/stirling/obj_tools/elf_reader.cc:170]: Nice. Thanks for the explanation. :)

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (oazizi000) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:95]: On second thought, since BaseOpenSSLTraceTest is the only test class in this test, you don't really need to reset the value on the destructor, because you always explicitly set the value during SetUp(), which is safe.

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (oazizi000) [inline src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:185]: Clever. Though also wonder if there's a better way. Apparently, we can't mix typed parameters with value parameters in gtest, which is unfortunate. So this might be the best way.

In theory, you might be able to turn the second value as a second type parameters, such that we `OpenSSLServerImplementations` gets converted to a sequence of `std::pairs`(like https://stackoverflow.com/questions/52256953/c-multiple-parameters-with-gtest-typed-test).

You can feel free to leave this as is, though, since it may not be worth the effort to change this.

**Review by oazizi000 — COMMENTED** [review]
This looks great @ddelnano. A few final comments, but overall, I have no objections. Hoping we can get this in soon.

Sorry for the tardy reviews on this!

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:185]: I tried something similar to that, but ran into difficulties so I'd prefer to leave this as is.

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:95]: But isn't it possible that any other 'trace bpf' tests that use uprobes could run later and they would have whatever the last value that was set be enabled? I thought that had the potential to pollute future tests, so I erred on the side of caution.

File: src/common/system/proc_parser.cc
Comment (ddelnano) [reply src/common/system/proc_parser.cc]: This was refactored in a032909130d1246d66d6fbb1adadabc6ee7b277e. I enhanced `ProcParser::ProcessSMaps` to work for my use case and consolidated the proc_parser function to work for both.

**Comment by ddelnano** [comment]
@oazizi000 the most pressing issue (refactoring the proc_parser code) is complete. The rest I was going to hold off on. Let me know what you think.

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (oazizi000) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:95]: There's no risk of that because each test is compiled into its own binary. So tests can't interfere with each other. The only risk would be if a new test class was added to `openssl_trace_bpf_test.cc`.

Anyways, not a big deal either way.

**Review by oazizi000 — APPROVED** [review]
Left a comment, but approved! Will look into pulling this in when I get a chance. As part of that, we'll run some perf regression tests to make sure it doesn't break anything performance-wise.

File: src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/openssl_trace_bpf_test.cc:95]: ah, I see. Since this is an easy fix, I've pushed another commit and retested to remove the `TearDown` method. Thanks for explaining that!

**Comment by oazizi000** [comment]
@ddelnano Can you rebase your changes onto main and squash them all into one commit? Trying to merge this in, but hitting merge conflicts.

**Comment by ddelnano** [comment]
I'm not confident that squashing and rebasing this branch (since I merged main in previously) would work well. So I opened up #428 with this branch's changes.

**Comment by oazizi000** [comment]
Replaced by #428.

**Pull request closed without merging**
